*** win/tclWinSock.c.ORG	Mon Jul 22 10:15:38 1996
--- win/tclWinSock.c	Mon Jul 22 10:15:46 1996
***************
*** 702,708 ****
  
      chan = Tcl_CreateChannel(&tcpChannelType, channelName, infoPtr->file,
  	    infoPtr->file, (ClientData) infoPtr);
!     if (Tcl_SetChannelOption(interp, chan, "-translation", "auto crlf") ==
              TCL_ERROR) {
          Tcl_Close((Tcl_Interp *) NULL, chan);
          return (Tcl_Channel) NULL;
--- 702,768 ----
  
      chan = Tcl_CreateChannel(&tcpChannelType, channelName, infoPtr->file,
  	    infoPtr->file, (ClientData) infoPtr);
!     if (Tcl_SetChannelOption(NULL, chan, "-translation", "auto crlf") ==
!             TCL_ERROR) { 
!         Tcl_Close((Tcl_Interp *) NULL, chan);
!         return (Tcl_Channel) NULL;
!     }
!     if (Tcl_SetChannelOption(NULL, chan, "-eofchar", "") == TCL_ERROR) {
!         Tcl_Close((Tcl_Interp *) NULL, chan);
!         return (Tcl_Channel) NULL;
!     }
!     return chan;
! }
! 
! /*
!  *----------------------------------------------------------------------
!  *
!  * Tcl_MakeTcpClientChannel --
!  *
!  *	Creates a Tcl_Channel from an existing client TCP socket.
!  *
!  * Results:
!  *	The Tcl_Channel wrapped around the preexisting TCP socket.
!  *
!  * Side effects:
!  *	None.
!  *
!  *----------------------------------------------------------------------
!  */
! 
! Tcl_Channel
! Tcl_MakeTcpClientChannel(sock)
!     ClientData sock;		/* The socket to wrap up into a channel. */
! {
!     SocketInfo *infoPtr;
!     char channelName[20];
!     Tcl_Channel chan;
!     int flag = 1;
! 
!     if (TclHasSockets(NULL) != TCL_OK) {
! 	return NULL;
!     }
! 
!     /*
!      * Set kernel space buffering and non-blocking.
!      */
! 
!     TclSockMinimumBuffers((SOCKET) sock, TCP_BUFFER_SIZE);
! 
!     if ((*winSock.ioctlsocket)((SOCKET)sock, FIONBIO, &flag) == SOCKET_ERROR) {
! 	TclWinConvertWSAError ((*winSock.WSAGetLastError)());
!         return NULL;
!     }
! 
!     infoPtr = NewSocketInfo (Tcl_GetFile((ClientData) sock,
!                                          TCL_WIN_SOCKET));
! 
!     sprintf(channelName, "sock%d", infoPtr->socket);
! 
!     chan = Tcl_CreateChannel(&tcpChannelType, channelName, infoPtr->file,
! 	    infoPtr->file, (ClientData) infoPtr);
!     if (Tcl_SetChannelOption((Tcl_Interp *) NULL, chan,
!                              "-translation", "auto crlf") ==
              TCL_ERROR) {
          Tcl_Close((Tcl_Interp *) NULL, chan);
          return (Tcl_Channel) NULL;
